byterun/ints.c

```diff
@@ -828,3 +849,308 @@ CAMLprim value caml_nativeint_of_string(value s)
+  custom_finalize_default,
+  uint32_cmp,
+  uint32_hash,
+  uint32_serialize,
+  uint32_deserialize, 
+  custom_compare_ext_default
+};
+
+CAMLprim value caml_copy_uint32(uint32_t u)
+{
+  value res = caml_alloc_custom(&caml_uint32_ops, 4, 0, 1);
+  Uint32_val(res) = u;
+  return res;
+}
+
+CAMLprim value caml_uint32_add(value a, value b)
+{ return caml_copy_uint32(Uint32_val(a) + Uint32_val(b)); }
+
+CAMLprim value caml_uint32_sub(value a, value b)
+{ return caml_copy_uint32(Uint32_val(a) - Uint32_val(b)); }
+
+CAMLprim value caml_uint32_mul(value a, value b)
+{ return caml_copy_uint32(Uint32_val(a) * Uint32_val(b)); }
+
+CAMLprim value caml_uint32_div(value a, value b)
+{ uint32_t dividend = Uint32_val(a);
+  uint32_t divisor = Uint32_val(b);
+  if (divisor == 0) caml_raise_zero_divide();
+  return caml_copy_uint32(dividend / divisor);
+}
+
+CAMLprim value caml_uint32_mod(value a, value b)
+{
+  uint32_t dividend = Uint32_val(a);
+  uint32_t divisor = Uint32_val(b);
+  if (divisor == 0) caml_raise_zero_divide();
+  return caml_copy_uint32(dividend % divisor);
+}
+
+CAMLprim value caml_uint32_and(value a, value b)
+{ return caml_copy_uint32(Uint32_val(a) & Uint32_val(b)); }
+
+CAMLprim value caml_uint32_or(value a, value b)
+{ return caml_copy_uint32(Uint32_val(a) | Uint32_val(b)); }
+
+CAMLprim value caml_uint32_xor(value a, value b)
+{ return caml_copy_uint32(Uint32_val(a) ^ Uint32_val(b)); }
+
+CAMLprim value caml_uint32_shift_left(value a, value b)
+{ return caml_copy_uint32(Uint32_val(a) << Int_val(b)); }
+
+CAMLprim value caml_uint32_shift_right(value a, value b)
+{ return caml_copy_uint32(Uint32_val(a) >> Int_val(b)); }
+
+CAMLprim value caml_uint32_of_int(value a)
+{ return caml_copy_uint32 (Long_val(a)); }
+
+CAMLprim value caml_uint32_to_int(value a)
+{ return Val_long(Uint32_val(a)); }
+
+CAMLprim value caml_uint32_of_int32(value a)
+{ return caml_copy_uint32 (Int32_val(a)); }
+
+CAMLprim value caml_uint32_to_int32(value a)
+{ return caml_copy_int32 (Uint32_val(a)); }
+
+CAMLprim value caml_uint32_of_string(value s)
+{
+  char * p;
+  uint32_t res, threshold;
+  int base, d;
+
+  p = parse_base(String_val(s), &base);
+  threshold = ((uint32_t) -1) / base;
+  d = parse_digit(*p);
+  if (d < 0 || d >= base) caml_failwith(UINT32_ERRMSG);
+  res = d;
+  for (p++; /*nothing*/; p++) {
+    char c = *p;
+    if (c == '_') continue;
+    d = parse_digit(c);
+    if (d < 0 || d >= base) break;
+    /* Detect overflow in multiplication base * res */
+    if (res > threshold) caml_failwith(UINT32_ERRMSG);
+    res = base * res + d;
+    /* Detect overflow in addition (base * res) + d */
+    if (res < (uint32_t) d) caml_failwith(UINT32_ERRMSG);
+  }
+  if (p != String_val(s) + caml_string_length(s)){
+    caml_failwith(UINT32_ERRMSG);
+  }
+  return caml_copy_uint32(res);
+}
+
+CAMLprim value caml_uint32_format(value fmt, value arg)
+{
+  char format_string[FORMAT_BUFFER_SIZE];
+
+  parse_format(fmt, ARCH_INT32_PRINTF_FORMAT, format_string);
```

Comment:
I wondered if using `ARCH_INT32_PRINTF_FORMAT` is correct here. This becomes `"l"`, which the C documentation says work for both signed and unsigned numbers. Maybe having a `ARCH_UINT32_PRINTF_FORMAT` alias would have been clearer, but I think the current code is fine.